### PR TITLE
 Refactor body of dbuf_free_range in to multiple functions

### PR DESCRIPTION
**Pull request by mattmacy**
async COW part 5

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [ ] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [ ] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [ ] I have updated the documentation accordingly.
- [ ] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [ ] I have run the ZFS Test Suite with this change applied.
- [ ] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

**Comment by ahrens** [comment]
Regarding the "doxygen" style comments: although to my eye these are ugly, I'm not totally opposed if you can show us the benefit.  What tools do you recommend for processing these comments?  Do we plan to publish the processed comments (e.g. on the website)?  Who would be the target audience? Maybe developers who would rather google it than bring up the source code?

File: module/zfs/dbuf.c [hunk 1]
```diff
@@ -302,6 +303,12 @@ dbuf_hash(void *os, uint64_t obj, uint8_t lvl, uint64_t blkid)
 	return (cityhash4((uintptr_t)os, obj, (uint64_t)lvl, blkid));
 }
 
+#define	DBUF_STATE_CHANGE(db, op, state, why) do {		\
```
Comment (wca) [inline module/zfs/dbuf.c]: This PR seems like it includes sections that are in #9923?  This is the original macro I was referring to.

Comment (mattmacy) [reply module/zfs/dbuf.c]: @wca It hasn’t been rebased.

**Comment by mattmacy** [comment]
> Regarding the "doxygen" style comments: although to my eye these are ugly, I'm not totally opposed if you can show us the benefit. What tools do you recommend for processing these comments? Do we plan to publish the processed comments (e.g. on the website)? Who would be the target audience? Maybe developers who would rather google it than bring up the source code?

@wca any thoughts on doxygen?

**Comment by mattmacy** [comment]
> Regarding the "doxygen" style comments: although to my eye these are ugly, I'm not totally opposed if you can show us the benefit. What tools do you recommend for processing these comments? Do we plan to publish the processed comments (e.g. on the website)? Who would be the target audience? Maybe developers who would rather google it than bring up the source code?

Doxygen use is pervasive within Linux. When looking at APIs or structures it's a much more succinct way of finding what one is looking for. The actual body of any implementation tends to be mostly get / set and bookkeeping which doesn't contribute to the purpose. You get the comments for all the key functions in one glance without having to wade through large amounts of code. That said, this is taken from Will's code and if he doesn't feel strongly enough about the value of doxygen comments to even chime after having been asked to do so, I'm perfectly willing to drop them if Brian also considers them "ugly".

**Comment by mattmacy** [comment]
@ahrens I removed the doxygen usage

**Comment by wca** [comment]
> @ahrens I removed the doxygen usage

I think this is probably the right path forward at this point in time.

In past efforts at upstreaming, Justin Gibbs and I proposed doxygenating ZFS, but the proposal failed to gain any traction, so we simply removed the doxygen annotations and left the contents as-is.  Which is unfortunate, because it rendered nicely browseable internal documentation.

https://illumos.topicbox.com/groups/zfs/T001593bdd1956600-M4fc599dd3924c1bea8b7776e/doxygenation-of-zfs-comments

File: module/zfs/dsl_crypt.c [hunk 2]
```diff
@@ -122,7 +122,7 @@ dsl_wrapping_key_create(uint8_t *wkeydata, zfs_keyformat_t keyformat,
 	/* allocate and initialize the underlying crypto key */
 	wkey->wk_key.ck_data = kmem_alloc(WRAPPING_KEY_LEN, KM_SLEEP);
 	if (!wkey->wk_key.ck_data) {
-		ret = ENOMEM;
+		ret = SET_ERROR(ENOMEM);
```
Comment (wca) [inline module/zfs/dsl_crypt.c]: This whole branch actually doesn't make sense, since the wait arg is `KM_SLEEP`.  It should just be removed.

Comment (mattmacy) [reply module/zfs/dsl_crypt.c]: @wca those sorts of mistakes are pervasive throughout dsl_crypt.c

Comment (mattmacy) [reply module/zfs/dsl_crypt.c]:
```
index 0635cc3c9..11e793240 100644
--- a/module/zfs/dsl_crypt.c
+++ b/module/zfs/dsl_crypt.c
@@ -116,15 +116,9 @@ dsl_wrapping_key_create(uint8_t *wkeydata, zfs_keyformat_t keyformat,
 
        /* allocate the wrapping key */
        wkey = kmem_alloc(sizeof (dsl_wrapping_key_t), KM_SLEEP);
-       if (!wkey)
-               return (SET_ERROR(ENOMEM));
 
        /* allocate and initialize the underlying crypto key */
        wkey->wk_key.ck_data = kmem_alloc(WRAPPING_KEY_LEN, KM_SLEEP);
-       if (!wkey->wk_key.ck_data) {
-               ret = SET_ERROR(ENOMEM);
-               goto error;
-       }
 
        wkey->wk_key.ck_format = CRYPTO_KEY_RAW;
        wkey->wk_key.ck_length = CRYPTO_BYTES2BITS(WRAPPING_KEY_LEN);
@@ -138,12 +132,6 @@ dsl_wrapping_key_create(uint8_t *wkeydata, zfs_keyformat_t keyformat,
 
        *wkey_out = wkey;
        return (0);
-
-error:
-       dsl_wrapping_key_free(wkey);
-
-       *wkey_out = NULL;
-       return (ret);
 }
 
 int
@@ -161,11 +149,6 @@ dsl_crypto_params_create_nvlist(dcp_cmd_t cmd, nvlist_t *props,
        char *keylocation = NULL;
 
        dcp = kmem_zalloc(sizeof (dsl_crypto_params_t), KM_SLEEP);
-       if (!dcp) {
-               ret = SET_ERROR(ENOMEM);
-               goto error;
-       }
-
        dcp->cp_cmd = cmd;
 
        /* get relevant arguments from the nvlists */
@@ -561,8 +544,6 @@ dsl_crypto_key_open(objset_t *mos, dsl_wrapping_key_t *wkey,
 
        /* allocate and initialize the key */
        dck = kmem_zalloc(sizeof (dsl_crypto_key_t), KM_SLEEP);
-       if (!dck)
-               return (SET_ERROR(ENOMEM));
 
        /* fetch all of the values we need from the ZAP */
        ret = zap_lookup(mos, dckobj, DSL_CRYPTO_KEY_CRYPTO_SUITE, 8, 1,
@@ -2430,9 +2411,7 @@ dsl_crypto_populate_key_nvlist(dsl_dataset_t *ds, uint64_t from_ivset_guid,
        VERIFY0(dmu_objset_from_ds(ds, &os));
        mdn = DMU_META_DNODE(os);
 
-       ret = nvlist_alloc(&nvl, NV_UNIQUE_NAME, KM_SLEEP);
-       if (ret != 0)
-               goto error;
+       nvlist_alloc(&nvl, NV_UNIQUE_NAME, KM_SLEEP);
 
        /* lookup values from the DSL Crypto Key */
        ret = zap_lookup(mos, dckobj, DSL_CRYPTO_KEY_CRYPTO_SUITE, 8, 1,
```
orthogonal to current work

File: module/zfs/dbuf.c [hunk 3]
```diff
@@ -1764,6 +1764,96 @@ dbuf_unoverride(dbuf_dirty_record_t *dr)
...
+		 */
+		dbuf_fix_old_data(db, txg);
+	}
+}
+
+
+static boolean_t
+dbuf_free_range_already_freed(dmu_buf_impl_t *db)
+{
+	/* XXX add comment about why these are OK */
```
Comment (behlendorf) [inline module/zfs/dbuf.c]: Then let's add that comment.

File: module/zfs/dbuf.c [hunk 4]
```diff
@@ -1764,6 +1764,96 @@ dbuf_unoverride(dbuf_dirty_record_t *dr)
...
+		 * contain new data when we sync.
+		 */
+		if (db->db_blkid != DMU_SPILL_BLKID &&
+		    db->db_blkid > dn->dn_maxblkid)
+			dn->dn_maxblkid = db->db_blkid;
+		dbuf_unoverride(dr);
+	} else {
+		/*
+		 * This dbuf is not dirty in the open context.
+		 * Either uncache it (if its not referenced in
```
Comment (behlendorf) [inline module/zfs/dbuf.c]: its -> it's

File: module/zfs/dbuf.c [hunk 5]
```diff
@@ -1764,6 +1764,96 @@ dbuf_unoverride(dbuf_dirty_record_t *dr)
...
+
+
+static boolean_t
+dbuf_free_range_already_freed(dmu_buf_impl_t *db)
+{
+	/* XXX add comment about why these are OK */
+	if (db->db_state == DB_UNCACHED || db->db_state == DB_NOFILL ||
+	    db->db_state == DB_EVICTING) {
+		ASSERT(db->db.db_data == NULL);
+		mutex_exit(&db->db_mtx);
```
Comment (behlendorf) [inline module/zfs/dbuf.c]: We should not be dropping the mutex in a different function than where it was originally taken.  This is at least commented, but my personal feeling in these changes make this logic more complicated.  I'd much rather we not add these helper functions (dbuf_free_range_already_freed, dbuf_free_range_filler_will_free, dbuf_clear_successful) and instead add a block comment explaining each case below.

File: module/zfs/dbuf.c [hunk 6]
```diff
@@ -1764,6 +1764,96 @@ dbuf_unoverride(dbuf_dirty_record_t *dr)
 }
 
 /*
+ * Disassociate the frontend for any older transaction groups of a
+ * dbuf that is inside a range being freed.  The primary purpose is to
+ * ensure that the state of any dirty records affected by the operation
+ * remain consistent.
+ */
+static void
+dbuf_free_range_disassociate_frontend(dmu_buf_impl_t *db, dnode_t *dn,
```
Comment (behlendorf) [inline module/zfs/dbuf.c]: I'm not sure what exactly "frontend" means here.  How about `dbuf_free_range_disassociate_dirty()` instead.

**Comment by behlendorf** [comment]
Closing for now.  Let's revisit this refactoring in the context of the full async COW PR.

**Pull request closed without merging**